File: indexes/cpp/src/kmeans.cpp
#include "kmeans.hpp"

#include <algorithm>
#include <cmath>
#include <limits>
#include <stdexcept>
#include <thread>

#include "distance.hpp"
#include "splitmix.hpp"

namespace vro {

std::size_t default_train_size(std::size_t n, std::size_t k) {
    return std::max(std::min(n, 256 * k), k);
}

std::size_t nearest_center(const float* point, const float* centers, std::size_t k,
                           std::size_t dim, Metric metric, float* score_out) {
    std::size_t best = 0;
    float best_score = -std::numeric_limits<float>::infinity();
    for (std::size_t c = 0; c < k; ++c) {
        const float* center = centers + c * dim;
        float s = metric == Metric::kDot ? dot(point, center, dim) : -l2_sq(point, center, dim);
        if (s > best_score) {
            best_score = s;
            best = c;
        }
    }
    if (score_out) *score_out = best_score;
    return best;
}

namespace {

// Step 2: k distinct training rows, drawn with next_below; a repeat draws again.
std::vector<float> init_centers(const float* points, std::size_t train_n, std::size_t dim,
                                std::size_t k, std::uint64_t seed) {
    SplitMix64 rng(seed);
    std::vector<bool> taken(train_n, false);
    std::vector<float> centers(k * dim);
    for (std::size_t c = 0; c < k; ++c) {
        std::size_t row;
        do {
            row = static_cast<std::size_t>(rng.next_below(train_n));
        } while (taken[row]);
        taken[row] = true;
        std::copy(points + row * dim, points + (row + 1) * dim, centers.begin() + c * dim);
    }
    return centers;
}

// Assigns every training row to its best center, in parallel over row ranges.
// Returns true if any assignment changed.
bool assign_all(const float* points, std::size_t train_n, std::size_t dim,
                const std::vector<float>& centers, std::size_t k, Metric metric, int threads,
                std::vector<std::int64_t>& assign, std::vector<float>& score) {
    std::size_t t_count = static_cast<std::size_t>(std::max(1, threads));
    t_count = std::min(t_count, std::max<std::size_t>(1, train_n));
    std::vector<char> changed(t_count, 0);
    auto work = [&](std::size_t t) {
        std::size_t begin = train_n * t / t_count;
        std::size_t end = train_n * (t + 1) / t_count;
        for (std::size_t i = begin; i < end; ++i) {
            float s = 0.0f;
            auto c = static_cast<std::int64_t>(
                nearest_center(points + i * dim, centers.data(), k, dim, metric, &s));
            if (c != assign[i]) changed[t] = 1;
            assign[i] = c;
            score[i] = s;
        }
    };
    std::vector<std::thread> pool;
    for (std::size_t t = 1; t < t_count; ++t) pool.emplace_back(work, t);
    work(0);
    for (auto& th : pool) th.join();
    return std::any_of(changed.begin(), changed.end(), [](char c) { return c != 0; });
}

void normalize_row(float* v, std::size_t dim) {
    float norm = std::sqrt(dot(v, v, dim));
    if (norm > 0.0f)
        for (std::size_t j = 0; j < dim; ++j) v[j] /= norm;
}

// Step (c): for each empty cluster in index order, move the worst-fit point
// (lowest score to its own center) into it. Only points whose cluster has at
// least 2 members qualify, so no cluster becomes empty by donating.
void fix_empty(std::size_t k, std::vector<std::int64_t>& labels, const std::vector<float>& score) {
    std::vector<std::size_t> counts(k, 0);
    for (auto l : labels) ++counts[static_cast<std::size_t>(l)];
    std::vector<bool> moved(labels.size(), false);
    for (std::size_t c = 0; c < k; ++c) {
        if (counts[c] != 0) continue;
        std::size_t worst = labels.size();
        for (std::size_t i = 0; i < labels.size(); ++i) {
            if (moved[i] || counts[static_cast<std::size_t>(labels[i])] < 2) continue;
            if (worst == labels.size() || score[i] < score[worst]) worst = i;
        }
        if (worst == labels.size()) break;  // cannot happen while train_n >= k
        --counts[static_cast<std::size_t>(labels[worst])];
        ++counts[c];
        labels[worst] = static_cast<std::int64_t>(c);
        moved[worst] = true;  // its old score no longer applies to its new cluster
    }
}

// Step (d): each center = mean of its points, then L2-normalized unless disabled.
void update_centers(const float* points, std::size_t dim, std::size_t k, bool normalize,
                    const std::vector<std::int64_t>& labels, std::vector<float>& centers) {
    std::vector<double> sums(k * dim, 0.0);
    std::vector<std::size_t> counts(k, 0);
    for (std::size_t i = 0; i < labels.size(); ++i) {
        auto c = static_cast<std::size_t>(labels[i]);
        ++counts[c];
        const float* p = points + i * dim;
        for (std::size_t j = 0; j < dim; ++j) sums[c * dim + j] += p[j];
    }
    for (std::size_t c = 0; c < k; ++c) {
        if (counts[c] == 0) continue;
        float* center = centers.data() + c * dim;
        for (std::size_t j = 0; j < dim; ++j)
            center[j] = static_cast<float>(sums[c * dim + j] / static_cast<double>(counts[c]));
        if (normalize) normalize_row(center, dim);
    }
}

}  // namespace

std::vector<float> kmeans(const float* points, std::size_t n, std::size_t dim,
                          const KMeansOptions& opt) {
    std::size_t k = opt.k;
    std::size_t train_n = opt.train_size ? std::min(opt.train_size, n) : default_train_size(n, k);
    train_n = std::min(std::max(train_n, k), n);
    if (k == 0 || train_n < k) throw std::invalid_argument("kmeans: need at least k training rows");

    std::vector<float> centers = init_centers(points, train_n, dim, k, opt.seed);
    std::vector<std::int64_t> assign(train_n, -1);
    std::vector<float> score(train_n, 0.0f);
    // Per iteration: (a) assign, (b) stop if no label changed since the
    // previous (a), (c) fix empty clusters, (d) recompute centers.
    for (int it = 0; it < opt.iters; ++it) {
        bool changed = assign_all(points, train_n, dim, centers, k, opt.metric, opt.threads,
                                  assign, score);
        if (!changed) break;
        std::vector<std::int64_t> labels = assign;  // (b) compares against unfixed labels
        fix_empty(k, labels, score);
        update_centers(points, dim, k, opt.normalize, labels, centers);
    }
    return centers;
}

}  // namespace vro

File: indexes/rust/src/kmeans.rs
//! The shared k-means of CONTRACT 6.2, used by ivf, pq, ivf_pq, and diskann.
//!
//! The assignment step runs in parallel with rayon. Call it inside
//! `ThreadPool::install` (as `main` does) to limit it to `--threads` threads.

use crate::distance::{dot, l2_sq};
use crate::splitmix::SplitMix64;
use rayon::prelude::*;

/// How points are assigned to centers.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Assign {
    /// Highest dot product (IVF centers, PQ with `metric=ip`).
    Dot,
    /// Lowest squared Euclidean distance (PQ with `metric=l2`).
    L2,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct KmeansOptions {
    pub assign: Assign,
    /// L2-normalize each center after the mean (step 3). PQ codebooks turn this off.
    pub normalize: bool,
}

impl Default for KmeansOptions {
    fn default() -> Self {
        Self {
            assign: Assign::Dot,
            normalize: true,
        }
    }
}

/// Runs k-means on `points` (row-major, `n` rows of `dim`), which are the training rows.
/// Returns the centers, row-major (k, dim).
pub fn kmeans(
    points: &[f32],
    dim: usize,
    k: usize,
    iters: usize,
    seed: u64,
    opts: KmeansOptions,
) -> Result<Vec<f32>, String> {
    let n = points.len() / dim;
    if k == 0 || n < k {
        return Err(format!(
            "k-means needs 1 <= k <= training rows, got k={k}, rows={n}"
        ));
    }
    let mut centers = init_centers(points, dim, k, seed);
    // Labels of step (a) of the previous iteration, before any empty-cluster fix.
    let mut prev: Vec<usize> = vec![usize::MAX; n];
    for _ in 0..iters {
        let (labels, fit) = assign_points(points, dim, &centers, opts.assign);
        if labels == prev {
            break;
        }
        prev = labels.clone();
        let mut labels = labels;
        centers = update_centers(points, dim, k, &mut labels, &fit, opts.normalize);
    }
    Ok(centers)
}

/// Step 2: k distinct training rows, drawn with `next_below(n)`; a repeat draws again.
fn init_centers(points: &[f32], dim: usize, k: usize, seed: u64) -> Vec<f32> {
    let n = points.len() / dim;
    let mut rng = SplitMix64::new(seed);
    let mut taken = vec![false; n];
    let mut centers = Vec::with_capacity(k * dim);
    while centers.len() < k * dim {
        let r = rng.next_below(n as u64) as usize;
        if !taken[r] {
            taken[r] = true;
            centers.extend_from_slice(&points[r * dim..(r + 1) * dim]);
        }
    }
    centers
}

/// The best center of `p`, and its fit (higher is better: dot, or minus the distance).
/// Ties go to the lower center index.
#[inline]
pub fn best_center(p: &[f32], centers: &[f32], dim: usize, assign: Assign) -> (usize, f32) {
    let mut best = (0usize, f32::NEG_INFINITY);
    for (c, center) in centers.chunks_exact(dim).enumerate() {
        let fit = match assign {
            Assign::Dot => dot(p, center),
            Assign::L2 => -l2_sq(p, center),
        };
        if fit > best.1 {
            best = (c, fit);
        }
    }
    best
}

/// Step 3, first half: each point's best center and fit, in parallel.
fn assign_points(
    points: &[f32],
    dim: usize,
    centers: &[f32],
    assign: Assign,
) -> (Vec<usize>, Vec<f32>) {
    points
        .par_chunks_exact(dim)
        .map(|p| best_center(p, centers, dim, assign))
        .unzip()
}

/// Step 3, second half, and step 4: means, the empty-cluster rule, and normalization.
fn update_centers(
    points: &[f32],
    dim: usize,
    k: usize,
    assign: &mut [usize],
    fit: &[f32],
    normalize: bool,
) -> Vec<f32> {
    let mut sums = vec![0.0f64; k * dim];
    let mut counts = vec![0usize; k];
    for (p, &c) in points.chunks_exact(dim).zip(assign.iter()) {
        counts[c] += 1;
        for (s, &x) in sums[c * dim..(c + 1) * dim].iter_mut().zip(p) {
            *s += x as f64;
        }
    }
    fix_empty_clusters(points, dim, assign, fit, &mut sums, &mut counts);
    let mut centers = vec![0.0f32; k * dim];
    for c in 0..k {
        let out = &mut centers[c * dim..(c + 1) * dim];
        let inv = 1.0 / counts[c].max(1) as f64;
        for (o, &s) in out.iter_mut().zip(&sums[c * dim..(c + 1) * dim]) {
            *o = (s * inv) as f32;
        }
        if normalize {
            normalize_in_place(out);
        }
    }
    centers
}

/// Step 4: each empty cluster, in index order, takes the worst-fit training point.
/// Only points whose cluster has at least 2 members are candidates, so no cluster
/// becomes empty by donating. Ties go to the lower point index.
fn fix_empty_clusters(
    points: &[f32],
    dim: usize,
    assign: &mut [usize],
    fit: &[f32],
    sums: &mut [f64],
    counts: &mut [usize],
) {
    let k = counts.len();
    for c in 0..k {
        if counts[c] > 0 {
            continue;
        }
        let worst = (0..assign.len())
            .filter(|&i| counts[assign[i]] > 1)
            .min_by(|&a, &b| fit[a].total_cmp(&fit[b]).then(a.cmp(&b)));
        let Some(i) = worst else { continue };
        let old = assign[i];
        let p = &points[i * dim..(i + 1) * dim];
        for (s, &x) in sums[old * dim..(old + 1) * dim].iter_mut().zip(p) {
            *s -= x as f64;
        }
        counts[old] -= 1;
        for (s, &x) in sums[c * dim..(c + 1) * dim].iter_mut().zip(p) {
            *s = x as f64;
        }
        counts[c] = 1;
        assign[i] = c;
    }
}

/// Scales `v` to length 1. A zero vector stays zero.
pub fn normalize_in_place(v: &mut [f32]) {
    let norm = dot(v, v).sqrt();
    if norm > 0.0 {
        v.iter_mut().for_each(|x| *x /= norm);
    }
}
